### ISSUE TITLE
Problem loading the project on Visual Studio

### ISSUE THREAD

**Issue by binte** [jmezach/MSBuild.Sdk.SqlProj#4]
I am having an issue using this tool. I have added the Project SDK and targetFramework property to a SQL Server project, as per your [guidelines ](https://github.com/jmezach/MSBuild.Sdk.SqlProj) and although it builds and generates the dacpac file on VS code and on the build server, I can't load the project on visual studio. I get the following error when I try to reload the project: 

<img width="418" alt="msbuild-sdk-error" src="https://user-images.githubusercontent.com/5520287/79390699-1b536d00-7f68-11ea-832e-8a214df959aa.png">

It seems to be a problem with some missing software on my machine, but I am also wondering if I am setting this up properly. Any hint on what I am doing wrong?

**Comment by jmezach** [comment]
That error looks like you don't have the .NET Core workload installed for Visual Studio. Make sure that workload is installed using the Visual Studio installer.

Also I was wondering which version of Visual Studio are you using?

**Comment by binte** [comment]
I do have, and I have just installed the remaining .NET Core and framework related workloads, just to be sure. The problem persists. 

I believe I can't target .NET standard on a SQL Server project. I am able to load, build and successfully generate a .dacpac file by:

- Creating a windows application project
- Copying the .sqlproj content from the database project to the newly created project's .csproj
- Adding the necessary changes in order to use MSBuild.Sdk.SqlProj, as well making sure that the output type is set to Database: 

`<OutputType>Database</OutputType>`

But it seems like a very big workaround to me... What do you recommend? 

I am using Visual Studio 2019 Enterprise btw

**Comment by jmezach** [comment]
Interesting. Would you mind sharing the contents of your .csproj here to help me figure this out?

**Comment by binte** [comment]
This is the .csproj in the project where I am able to build the solution and generate the .dacpac

```
<?xml version="1.0" encoding="utf-8"?>
<Project Sdk="MSBuild.Sdk.SqlProj/1.0.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="4.0">
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <Name>MyProject.Database</Name>
    <SchemaVersion>2.0</SchemaVersion>
    <ProjectVersion>4.1</ProjectVersion>
    <ProjectGuid>{5399f622-031d-48d6-a251-55deaca6ac65}</ProjectGuid>
    <DSP>Microsoft.Data.Tools.Schema.Sql.Sql140DatabaseSchemaProvider</DSP>
    <OutputType>Database</OutputType>
    <RootPath>
    </RootPath>
    <RootNamespace>MyProject.Database</RootNamespace>
    <AssemblyName>MyProject.Database</AssemblyName>
    <ModelCollation>1033, CI</ModelCollation>
    <DefaultFileStructure>BySchemaAndSchemaType</DefaultFileStructure>
    <DeployToDatabase>True</DeployToDatabase>
    <TargetFramework>netstandard2.0</TargetFramework>
    <TargetLanguage>CS</TargetLanguage>
    <AppDesignerFolder>Properties</AppDesignerFolder>
    <SqlServerVerification>False</SqlServerVerification>
    <IncludeCompositeObjects>True</IncludeCompositeObjects>
    <TargetDatabaseSet>True</TargetDatabaseSet>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <OutputPath>bin\Release\</OutputPath>
    <BuildScriptName>$(MSBuildProjectName).sql</BuildScriptName>
    <TreatWarningsAsErrors>False</TreatWarningsAsErrors>
    <DebugType>pdbonly</DebugType>
    <Optimize>true</Optimize>
    <DefineDebug>false</DefineDebug>
    <DefineTrace>true</DefineTrace>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <OutputPath>bin\Debug\</OutputPath>
    <BuildScriptName>$(MSBuildProjectName).sql</BuildScriptName>
    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <DefineDebug>true</DefineDebug>
    <DefineTrace>true</DefineTrace>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup>
    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">16.0</VisualStudioVersion>
    <SSDTExists Condition="Exists('$(MSBuildExtensionsPath)\Microsoft\VisualStudio\v$(VisualStudioVersion)\SSDT\Microsoft.Data.Tools.Schema.SqlTasks.targets')">True</SSDTExists>
    <VisualStudioVersion Condition="'$(SSDTExists)' == ''">16.0</VisualStudioVersion>
  </PropertyGroup>
  <ItemGroup>
    <Folder Include="Properties" />
    <Folder Include="dbo" />
    <Folder Include="dbo\Tables" />
  </ItemGroup>
  <ItemGroup>
    <Build Include="**\*.sql" />
  </ItemGroup>
</Project>
```

**Comment by ErikEJ** [comment]
But that is not a SDK project !?

**Comment by jmezach** [comment]
Indeed what @ErikEJ says, this is not an SDK-style project file. Try to bring it back to just what I specified in the [README](https://github.com/jmezach/MSBuild.Sdk.SqlProj/blob/master/README.md). You should be able to open the project in Visual Studio then.

Looking at your project file you posted, you might want to add a `<SqlServerVersion>140</SqlServerVersion>` element to a `<PropertyGroup>` to make sure you're targeting SQL Server 2017. Then have a look at the [Microsoft documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.dac.model.tsqlmodeloptions?view=sql-dacfx-140.3881.1) to see if there are any additional properties you might want to set.

**Comment by binte** [comment]
It worked, thank you. I was not sure what was an SDK project, so I ended up mixing it up with the SQL Server project. 

For me it was a requirement to be able to use SQL Server Object Explorer to update the database, and I think that's only possible with SQL Server (.sqlproj) projects. I managed to overcome this by having a SQL Server project with the SQL database scripts, and within a pre-build event in the SDK project I copy the scripts to the SDK project, and I remove them in a post-build event. 

Then when I build, I ignore the SQL Server project

**Comment by jmezach** [comment]
Good to hear that it is working for you now. And thank you for trying it out. All feedback is welcome.

That actually sounds like an interesting trick to keep the SSDT features, while also benefitting from MSBuild.Sdk.SqlProj. Would you mind sharing your solution? Either here or through a pull request to the docs? That would be awesome.

**Comment by ErikEJ** [comment]
It should poitentially be possible to simply use linked files?

**Comment by binte** [comment]
@jmezach 

I have some more feedback. 

On my CD pipeline, after building the SDK project and getting the .dacpac file, I am comparing it to a live DB and generating a schema changes script that, after validation, I will later deploy to update the live DB. But no changes were ever being generated, even though that I was updating the .SQL files.

When I locally built the SQL Server project on Visual Studio (the SQL Server project fails to build on the pipeline and on VS code because of the missing targets, but it does build on Visual Studio) and then used the generated .dacpac file to generate the schema changes script locally with `sqlpackage.exe`, I finally got some changes on the generated script. 

The two generated files should have been identical. Thus I thought my configuration on the SDK project could be the root of this weird behaviour and I removed the pre-build event and added the .SQL scripts manually, before building the SDK project again. And after running `sqlpackage.exe /action:script` with this newly generated file I got a valid  schema changes script again.

I didn't open a new issue because this is clearly a problem on my end. My guess is, the files are being created and deleted so fast, that the build is not actually picking its content. 

I have tried @ErikEJ's suggestion yesterday unsuccessfully, but I will give it a try again later today.

**Comment by jmezach** [comment]
@binte Good to hear that you managed to fix it. Obviously the files need to be there at the time of the build for MSBuild to pick them up.

I'm wondering how you've tried to link the files? I would think that if you've put the SQL files in the folder containing your .csproj file you could create a separate folder containing a .sqlproj and include all the files in there from the folder containing the .csproj. In fact, it may even work if you have both the .sqlproj and the .csproj in the same folder.

Haven't tried that though because we're not actually using much of the SSDT stuff in our projects to be honest. But do let me know if that works, because it's worth adding that to the docs.

**Comment by binte** [comment]
I tried several different ways to link the files. Whatever I did that implied the files to be in the SQL Server project physically, and then linked on the SDK project, these would not be picked up during the build. Just as expected! 

In fact, even the solution I previously suggested, by having the pre-build and post-build events, does not work. Probably there is a way to do it with more advanced commands, but whatever I tried, if the files would not physically be in some folder inside the SDK project before the build started, these would not be picked up. 

I tried your last suggestion as well, but that would easily break the link with the SQL Server Object Explorer. 

I decided to stick with a simple PowerShell routine on the CD pipeline that copies the .sql files from the SQL Server project to the SDK project. This is more stable than any of the previous solutions, I can fully use the SQL Server Object Explorer locally, my local builds don't break neither on Visual Studio nor on VS Code, and I finally have the CD pipeline up and running. The only problem is that if I want to build the SDK project locally with the current .sql files I would have to manually copy them. Not sexy, but I can live with that.

Thank you for the support!
